File: 프로그래머스/lv3/49189. 가장 먼 노드/가장 먼 노드.py
from collections import deque
def solution(n, edge):
    answer = 0
    que = deque()
    edge.sort()
    distance = [0]*(n+1)
    route = [[]for i in range(n+1)]
    for i in edge:
        route[i[0]].append(i[1])
        route[i[1]].append(i[0])
    que.append(1)
    while que:
        cur = que.popleft()
        for i in route[cur]:
            if distance[i] == 0:
                que.append(i)
                distance[i] = distance[cur]+1
    distance[1] = 0
    max_dis = max(distance)
    for i in distance:
        if i == max_dis:
            answer += 1
            
    return answer

File: 프로그래머스/3/49189. 가장 먼 노드/가장 먼 노드.java
import java.util.*;

class Solution {
    
    static boolean[][] map;
    static boolean[] visit;
    static int[] dis = new int[50001];
    static class Node{
        int node;
        int dis;
        public Node(int node, int dis){
            this.node = node;
            this.dis = dis;
        }
    }
    static int max_dis = 0;
    public int solution(int n, int[][] edge) {
        int answer = 0;
        map = new boolean[n + 1][n + 1];
        visit = new boolean[n + 1];
        for(int[] e : edge){
            //간선 처리
            map[e[0]][e[1]] = true;
            map[e[1]][e[0]] = true;
        }
        visit[1] = true;
        Queue<Node> q = new ArrayDeque<>();
        q.offer(new Node(1, 0));
        while(!q.isEmpty()){
            Node cur = q.poll();
            for(int i = 1; i <= n; i++){
                //자신 또는 방문한 노드라면 건너뛰기
                if(i == cur.node || visit[i]) continue;
                if(map[cur.node][i]){
                    //갈 수 있다면
                    q.offer(new Node(i, cur.dis + 1));
                    dis[cur.dis + 1] ++;
                    max_dis = Math.max(max_dis, cur.dis + 1);
                    visit[i] = true;
                }
            }
        }
        answer = dis[max_dis];
        return answer;
    }
}